### PR TITLE
mpich: fix the detection of the package

### PR DESCRIPTION
**Pull request by skosukhin**
This mainly addresses problems on `darwin` with MacPorts-built `mpich`.

1. Fix `Warning: error detecting "mpich" from prefix /opt/local/bin [expected str, bytes or os.PathLike object, not NoneType]` by checking whether `spack_compiler.cc` is set before using its value (`Spack` falsely detects `apple-clang` in `/opt/local/bin` that is mixed with `gfortran` but does not have `cc`).
2. Fix the detection of MacPorts-built `gcc`, which has suffix `-mp-10` on my system.
3. Fix incorrect concatenation of a multi-valued variant with a compiler spec (`Warning: error detecting "mpich" from prefix /opt/local/bin [invalid values for variant "device" in package "mpich": ['ch4%gcc@10.3.0']`)

### CODE REVIEW

**Review by alalazo — APPROVED** [review]
Waiting to receive some comments from maintainers before merging

**Review by raffenet — APPROVED** [review]
LGTM

**Pull request merged**
